### ISSUE TITLE
Discovery onEndpointLost never called?

### ISSUE THREAD

**Issue by butchmarshall** [googlesamples/android-nearby#49]
`onEndpointFound` is successfully called, but `onEndpointLost` never receives an event for some reason.

If I kill the advertiser and restart it, I then get duplicate keys inserted into `onEndpointFound`.

Basically, I'm always stuck with my "discovered" endpoints even when they stop advertising and/or go out of range.

```
clientSingleton.startDiscovery(
	getServiceId(),
	new EndpointDiscoveryCallback() {
		@Override
		public void onEndpointFound(String endpointId, DiscoveredEndpointInfo info) {
			if (serviceId.equals(info.getServiceId())) {
				Endpoint endpoint = new Endpoint(endpointId, info.getEndpointName());
				mDiscoveredEndpoints.put(endpointId, endpoint);
				onEndpointDiscovered(endpoint);
			}
		}

		@Override
		public void onEndpointLost(String endpointId) {
			//Log.d("NearbyConnectionModule", "ENDPOINT LOST!");
			onEndpointLost(endpointId);
		}
	}, discoveryOptions)
.addOnSuccessListener(
	new OnSuccessListener<Void>() {
		@Override
		public void onSuccess(Void unusedResult) {
			onDiscoveryStarted();
		}
	}
).addOnFailureListener(
	new OnFailureListener() {
		@Override
		public void onFailure(@NonNull Exception e) {
			mIsDiscovering = false;
			onDiscoveryStartFailed(e.getMessage());
		}
	}
);
```

**Comment by alexkang** [comment]
Hi! What version of Google Play Services are your devices running?

**Comment by butchmarshall** [comment]
Hi @AlexKang  thanks for replying!

I did notice that this example app uses `compile 'com.google.android.gms:play-services-nearby:14.0.0'`

I cannot get v14 installed.  My android studio has everything possible downloaded.

![image](https://user-images.githubusercontent.com/339763/35424557-8118aef4-0222-11e8-9722-865cb90d90ed.png)

![image](https://user-images.githubusercontent.com/339763/35424698-52031ea0-0223-11e8-9bc7-e12e34c82d00.png)


top build.gradle

```
buildscript {
    repositories {
        jcenter()
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:2.3.3'
        classpath 'com.google.gms:google-services:3.1.2'

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        mavenLocal()
        jcenter()
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            url "$rootDir/../node_modules/react-native/android"
        }
        maven { url "https://maven.google.com" }
    }
}
```

my app build.gradle
```

android {
    compileSdkVersion 25
    buildToolsVersion "25.0.3"

    defaultConfig {
        applicationId "com.googlenearbyconnection"
        minSdkVersion 16
        targetSdkVersion 25
        versionCode 1
        versionName "1.0"
        ndk {
            abiFilters "armeabi-v7a", "x86"
        }
        multiDexEnabled true
    }
    splits {
        abi {
            reset()
            enable enableSeparateBuildPerCPUArchitecture
            universalApk false  // If true, also generate a universal APK
            include "armeabi-v7a", "x86"
        }
    }
    buildTypes {
        release {
            minifyEnabled enableProguardInReleaseBuilds
            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
        }
    }
    // applicationVariants are e.g. debug, release
    applicationVariants.all { variant ->
        variant.outputs.each { output ->
            // For each separate APK per architecture, set a unique version code as described here:
            // http://tools.android.com/tech-docs/new-build-system/user-guide/apk-splits
            def versionCodes = ["armeabi-v7a":1, "x86":2]
            def abi = output.getFilter(OutputFile.ABI)
            if (abi != null) {  // null for the universal-debug, universal-release variants
                output.versionCodeOverride =
                        versionCodes.get(abi) * 1048576 + defaultConfig.versionCode
            }
        }
    }
}

repositories {
    mavenCentral()
}

dependencies {
    compile project(':react-native-vector-icons')
    compile fileTree(dir: "libs", include: ["*.jar"])
    compile "com.facebook.react:react-native:+"  // From node_modules
    compile 'com.google.android.gms:play-services-nearby:11.8.0'
    compile 'com.android.support:appcompat-v7:25.3.1'
    compile 'com.android.support:multidex:1.0.1'
}

```
When I attempt 14.0.0

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':app'.
> Could not resolve all dependencies for configuration ':app:_debugApkCopy'.
   > Could not find com.google.android.gms:play-services-nearby:14.0.0.
     Searched in the following locations:
         file:/home/butch/Android/Sdk/extras/google/m2repository/com/google/android/gms/play-services-nearby/14.0.0/play-services-nearby-14.0.0.pom
         file:/home/butch/Android/Sdk/extras/google/m2repository/com/google/android/gms/play-services-nearby/14.0.0/play-services-nearby-14.0.0.jar
         file:/home/butch/workspace/GoogleNearbyConnection/android/sdk-manager/com/google/android/gms/play-services-nearby/14.0.0/play-services-nearby-14.0.0.jar
         file:/home/butch/Android/Sdk/extras/android/m2repository/com/google/android/gms/play-services-nearby/14.0.0/play-services-nearby-14.0.0.pom
         file:/home/butch/Android/Sdk/extras/android/m2repository/com/google/android/gms/play-services-nearby/14.0.0/play-services-nearby-14.0.0.jar
         file:/home/butch/workspace/GoogleNearbyConnection/android/sdk-manager/com/google/android/gms/play-services-nearby/14.0.0/play-services-nearby-14.0.0.jar
     Required by:
         project :app
```

```
/home/butch/Android/Sdk/extras/google/m2repository/com/google/android/gms/play-services-nearby/
10.0.0  10.2.0  10.2.4  11.0.0  11.0.2  7.0.0  7.5.0  8.1.0  8.4.0  9.0.1  9.2.0  9.4.0  9.6.1  maven-metadata.xml      maven-metadata.xml.sha1
10.0.1  10.2.1  10.2.6  11.0.1  11.0.4  7.3.0  7.8.0  8.3.0  9.0.0  9.0.2  9.2.1  9.6.0  9.8.0  maven-metadata.xml.md5
```

**Comment by butchmarshall** [comment]
I have open sourced the wrapper library i'm building around this for react native.   Only this issue remains. 

https://www.npmjs.com/package/react-native-google-nearby-connection

**Comment by Xlythe** [comment]
That's my bad, I'll fix the sample apps / docs. Use the latest play-services version (11.8.1, I believe).

**Comment by alexkang** [comment]
Hey! Sorry for the late reply. So to recreate this, do these steps seem appropriate:

1. Device A advertises.
2. Device B discovers.
3. Device B discovers Device A.
4. Device A stops advertising.
5. Wait some amount of time?
6. Device A starts advertising again.
7. Device B gets a second onEndpointFound?

And how consistently does this occur?

**Comment by butchmarshall** [comment]
Hi @AlexKang - you got it - except # 7 never occurs, and device B can connect once again to A after # 6 (its like it never stopped advertising).

I've almost got an example application ready to go along with my wrapper library [react-native-google-nearby-connection](https://www.npmjs.com/package/react-native-google-nearby-connection) - this should make the problem pretty easy to reproduce.

**Comment by butchmarshall** [comment]
OK - the example application is there.  

Once you compile and run the example application tail the logs on your device with `adb logcat | grep NearbyConnection` you'll notice this log in `onEndpointLost` will never fire the log

`logD(String.format("onEndpointLost(endpointId=%s)", endpointId));`

**Comment by alexkang** [comment]
Thanks! I'll take a look at the example app today.

As for onEndpointLost, how long are you waiting for it? And does anything change if you wait a long amount of time (i.e. leave the devices there for ~3 minutes after one stops advertising)?

**Comment by alexkang** [comment]
Hi again! So I tried installing your [example app](https://github.com/butchmarshall/react-native-google-nearby-connection/tree/master/example), but I ran into this when doing `npm install`:
```
npm ERR! native-base@2.3.7 postinstall: `node addEjectScript.js`
npm ERR! Exit status 8
npm ERR! 
npm ERR! Failed at the native-base@2.3.7 postinstall script 'node addEjectScript.js'.
```
Sorry, I'm not too familiar with React. I was able to get onEndpointLost triggered however with the test apps I'm using.

This should help narrow down the issue:
1. What devices are you using and what version of Android are they on?
2. The question from my previous post. Has waiting ~3 minutes changed anything?

Thanks!

**Comment by butchmarshall** [comment]
NativeBase wasn't actually used in the current master example - i've removed it from package.json and pushed an update.

Mind giving installing another try?  Ensure you're on master branch and it *should* work for you now (I'm under linux, maybe you're windows/mac?)

1) Having this issue on Samsung S7 Edge, S4, and S3.
2) Yep - waited as long as 20 with no change

**Comment by alexkang** [comment]
Thanks! I was able to get your app to build and install on my devices. And I think I found the issue!

So take a look at your [wrapper code here](https://github.com/butchmarshall/react-native-google-nearby-connection/blob/master/android/src/main/java/com/butchmarshall/reactnative/google/nearby/connection/NearbyConnectionModule.java#L1018). You call `stopDiscovering` when an endpoint is discovered. However, onLost reporting only works as long as discovery is active.

Let me know if you have any further questions!
